Add support for using \input with absolute paths #254

Lukas-Heiligenbrunner opened the pull request:
<!-- The issue that is fixed by this PR, if applicable: -->
Fixes #254, fixes #1243

#### Summary of additions and changes

* fixed path completion for relative as well as absolute paths. 
* added wrapper class `LatexPathProviderBase` to represent all possible completions
  * only folder completion
  * only file completion
  * file and folder completion
  * include graphicspath in completion?
* in \graphicspath command only folder completion
* in \includegraphics include also graphicspath in completion
* match valid extensions defined in `LatexRegularCommand`
#### How to test this pull request
following sketch:
```
\documentclass[11pt]{article}
\usepackage{graphicx}

\graphicspath{{./myfolder/}}

\begin{document}
    \input{/home/lukas/IdeaProjects/Bla/help.tex}

    \includegraphics{cpplogo.png}
    \includegraphics{/home/lukas/IdeaProjects/Bla/rsc/cpplogo.png}
    \includegraphics{cpplogo2.png}
\end{document}

```

Lukas-Heiligenbrunner commented:
This error has to do with the `LatexNoExtensionInspection`. The inspection forces to remove the .tex extension and then the file can´t be found anymore. Why is there even an error thrown? The path with extension is also valid, isn´t it? Maybe we can replace it with a weak warning? 
Moreover, why is this error thrown with a `\include` only and not with a `\input`. Is there any difference in path resolution? 

However, trying to add omitted default extensions to file inspection now and add some test cases.

PHPirates commented:
>This error has to do with the LatexNoExtensionInspection. The inspection forces to remove the .tex extension and then the file can´t be found anymore.

If I understand you correctly, you are wondering why `\include{somefile.tex}` is wrong. You can read about it here: https://tex.stackexchange.com/a/319512/98850

I don't know exactly if there is a difference for \include or \input (or other include commands) regarding absolute paths, but let's try to keep as close to LaTeX's behaviour as possible, with the difference that we should give clearer warning messages than pdflatex does :)

PHPirates commented:
I just tried \include with an absolute path and it didn't work, maybe I'm doing something wrong or \include doesn't support it (in which case we perhaps should show a different error message than a confusing 'file not found').

At least for \graphicspath and \input absolute paths work, maybe you can test the other include commands.

Lukas-Heiligenbrunner commented:
> I just tried \include with an absolute path and it didn't work, maybe I'm doing something wrong or \include doesn't support it (in which case we perhaps should show a different error message than a confusing 'file not found').
> 
> At least for \graphicspath and \input absolute paths work, maybe you can test the other include commands.

That´s correct `\include` doesn´t support that. I´ve also tested with `\includeonly` and it didn´t work  with absolute paths as well. 
Maybe there should be a new inspection called sth. like: `No absolute path allowed here` with corresponding commands defined in `Magic.kt`?

PHPirates commented:
> Maybe there should be a new inspection called sth. like: No absolute path allowed here with corresponding commands defined in Magic.kt?

That could certainly avoid some future confusion, it might also help if we don't provide absolute path autocompletion in that command, but I don't know if that's easy to do with the current implementation.

Lukas-Heiligenbrunner commented:
> That could certainly avoid some future confusion, it might also help if we don't provide absolute path autocompletion in that command, but I don't know if that's easy to do with the current implementation.

Fine, I'll also do that. Should't be that hard to implement.

Lukas-Heiligenbrunner commented:
An error is thrown on \include and \includeonly with absolute paths now. 
The completion restriction should also work now. 

But I have still some problems with the test cases:
I tested:
```
    @Test
    fun testCompleteRelativePath(){
        myFixture.configureByText("main.tex","""\input{<caret>}""")
        myFixture.configureByText("main2.tex","""bla txt ext""")

        val result = myFixture.complete(CompletionType.BASIC)

        assert(result.size > 0)
    }
```

My expected behaviour: autocompletion finds main2.tex and gives result.size==1
But actually it is always zero. What am I doing wrong? 
And how would be the best way to test absolute paths?

PHPirates commented:
I'll check tomorrow, but don't you need to pass LatexFileType to configureByText?
To test absolute paths, I guess you need to somehow find the absolute path to the test file or something.

Lukas-Heiligenbrunner commented:
> I'll check tomorrow, but don't you need to pass LatexFileType to configureByText?
> To test absolute paths, I guess you need to somehow find the absolute path to the test file or something.

I did that with the same result...

```
    @Test
    fun testCompleteRelativePath(){
        myFixture.configureByText(LatexFileType,"""\input{<caret>}""")
        myFixture.configureByText(LatexFileType,"""bla txt text""")

        val result = myFixture.complete(CompletionType.BASIC)

        assert(result.size > 0)
    }
```

Ok, thank you a lot!

PHPirates commented:
@Lukas-Heiligenbrunner Check the test I just pushed

Lukas-Heiligenbrunner commented:
> @Lukas-Heiligenbrunner Check the test I just pushed

Your test gives me an Assertion-Failed error. 
Did it work in your case?

PHPirates commented:
Strange, the test did work locally. But when I just tried to rewrite it, it didn't work anymore. 
I didn't manage to fix it, so let's just do without the tests for now and look at it later.

Lukas-Heiligenbrunner commented:
> Strange, the test did work locally. But when I just tried to rewrite it, it didn't work anymore.
> I didn't manage to fix it, so let's just do without the tests for now and look at it later.

I added some testcases now. (except relative path ones, don't know how they work)
can you still imagine additional testcases that I should implement?

PHPirates commented on src/nl/hannahsten/texifyidea/completion/pathcompletion/LatexGraphicsPathProvider.kt:
```diff
@@ -0,0 +1,43 @@
+                    .childrenOfType(LatexNormalText::class).forEach { path ->
+                        // Check if graphicspath is an absolute or relative path
+                        if (File(path.text).isAbsolute) {
+                            file.findRootFile().containingDirectory.virtualFile.fileSystem.findFileByPath(path.text)?.apply {
```
In all places where you added `file.findRootFile().containingDirectory` we need to protect against it's being null, see #1243. 
See the contract of getContainingDirectory(): `@return the containing directory, or {@code null} if the file exists only in memory.`

Lukas-Heiligenbrunner commented:
> * When I use an absolute graphicspath, the autocomplete in \includegraphics works correctly but the file not found inspection cannot find the file, even though it compiles correctly. With a relative path in \graphicspath it _can_ find it.

I overlooked, my mistake. Should work correctly now. 

> * I don't know whether it was already broken or not, but when I write \input{doesnotexist} and use the quickfix to create the file, the file is created without file extension. Same with an absolute path.

I did this on purpose, I thought it would be better to display what the user enters. Anyway, I changed it to autocomplete to the default extension now. 
> * When you add an inspection, you should add a description in resources/inspectionDescriptions, now when you Ctrl+F1 twice to get the long description it says 'under construction'.

Added. 
> * Perhaps you can add a quick explanation about the file not found/absolute path inspections at https://github.com/Hannah-Sten/TeXiFy-IDEA/wiki/Probable-bugs#File-not-found ?

Added.

Lukas-Heiligenbrunner commented:
#1149 
was it something like that you thought?
Any further suggestions?

Lukas-Heiligenbrunner commented:
No problem, you are welcome (:
